#  다중 클래스 분류기
- 3개 이상의 클래스(범주)가 있는 경우

In [1]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

In [2]:
import pandas as pd

In [3]:
### data load
iris = pd.read_csv("https://vincentarelbundock.github.io/Rdatasets/csv/datasets/iris.csv")
iris

Unnamed: 0  Sepal.Length  Sepal.Width  Petal.Length  Petal.Width  \
0             1           5.1          3.5           1.4          0.2   
1             2           4.9          3.0           1.4          0.2   
2             3           4.7          3.2           1.3          0.2   
3             4           4.6          3.1           1.5          0.2   
4             5           5.0          3.6           1.4          0.2   
..          ...           ...          ...           ...          ...   
145         146           6.7          3.0           5.2          2.3   
146         147           6.3          2.5           5.0          1.9   
147         148           6.5          3.0           5.2          2.0   
148         149           6.2          3.4           5.4          2.3   
149         150           5.9          3.0           5.1          1.8   

       Species  
0       setosa  
1       setosa  
2       setosa  
3       setosa  
4       setosa  
..         ...  
145  virginica  
146  virginica  
147  virginica  
148  virginica  
149  virginica  

[150 rows x 6 columns]

In [4]:
iris.loc[50:,]

Unnamed: 0  Sepal.Length  Sepal.Width  Petal.Length  Petal.Width  \
50           51           7.0          3.2           4.7          1.4   
51           52           6.4          3.2           4.5          1.5   
52           53           6.9          3.1           4.9          1.5   
53           54           5.5          2.3           4.0          1.3   
54           55           6.5          2.8           4.6          1.5   
..          ...           ...          ...           ...          ...   
145         146           6.7          3.0           5.2          2.3   
146         147           6.3          2.5           5.0          1.9   
147         148           6.5          3.0           5.2          2.0   
148         149           6.2          3.4           5.4          2.3   
149         150           5.9          3.0           5.1          1.8   

        Species  
50   versicolor  
51   versicolor  
52   versicolor  
53   versicolor  
54   versicolor  
..          ...  
145   virginica  
146   virginica  
147   virginica  
148   virginica  
149   virginica  

[100 rows x 6 columns]

In [6]:
### 설명변수(x)와 반응변수(y)
x = iris[["Sepal.Length","Sepal.Width","Petal.Length","Petal.Width"]]
y = iris["Species"]

In [7]:
### Train & Test data
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.4)

In [9]:
### logistic regression
# multi_class = {'auto','ovr','multinomial'}, default= 'auto'
logistic = LogisticRegression(multi_class = 'ovr')
model = logistic.fit(x_train,y_train)

In [11]:
### 예측
y_pred = model.predict(x_test)
y_pred

array(['versicolor', 'setosa', 'virginica', 'setosa', 'virginica',
       'setosa', 'versicolor', 'virginica', 'versicolor', 'versicolor',
       'versicolor', 'versicolor', 'setosa', 'virginica', 'setosa',
       'setosa', 'setosa', 'versicolor', 'virginica', 'virginica',
       'setosa', 'virginica', 'setosa', 'virginica', 'virginica',
       'setosa', 'virginica', 'setosa', 'virginica', 'setosa',
       'virginica', 'virginica', 'setosa', 'versicolor', 'virginica',
       'setosa', 'setosa', 'versicolor', 'setosa', 'versicolor',
       'virginica', 'virginica', 'setosa', 'versicolor', 'setosa',
       'versicolor', 'virginica', 'setosa', 'virginica', 'setosa',
       'versicolor', 'setosa', 'versicolor', 'setosa', 'setosa',
       'virginica', 'setosa', 'virginica', 'versicolor', 'virginica'],
      dtype=object)

In [12]:
### 교차표
pd.crosstab(y_test, y_pred) # 행이름 : y_test

col_0       setosa  versicolor  virginica
Species                                  
setosa          24           0          0
versicolor       0          15          2
virginica        0           0         19

In [15]:
from sklearn.metrics import classification_report

In [16]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

      setosa       1.00      1.00      1.00        24
  versicolor       1.00      0.88      0.94        17
   virginica       0.90      1.00      0.95        19

    accuracy                           0.97        60
   macro avg       0.97      0.96      0.96        60
weighted avg       0.97      0.97      0.97        60



###  house 가격 예측하기

In [22]:
# 설명변수 후보  bedrooms, bathrooms, sqft_living, view, condition,sqft_above, yr_built, yr_renovated
# 반응변수 price_cat


In [37]:
house = pd.read_csv('http://youngho.iwinv.net/data/house_price_prediction.csv')
house.head()

date     price price_cat  bedrooms  bathrooms  sqft_living  \
0  2014-05-02 0:00  313000.0    Class3         3       1.50         1340   
1  2014-05-02 0:00  342000.0    Class3         3       2.00         1930   
2  2014-05-02 0:00  420000.0    Class3         3       2.25         2000   
3  2014-05-02 0:00  550000.0    Class2         4       2.50         1940   
4  2014-05-02 0:00  490000.0    Class2         2       1.00          880   

   sqft_lot  floors  waterfront  view  condition  sqft_above  sqft_basement  \
0      7912     1.5           0     0          3        1340              0   
1     11947     1.0           0     0          4        1930              0   
2      8030     1.0           0     0          4        1000           1000   
3     10500     1.0           0     0          4        1140            800   
4      6380     1.0           0     0          3         880              0   

   yr_built  yr_renovated                    street       city  statezip  \
0      1955          2005      18810 Densmore Ave N  Shoreline  WA 98133   
1      1966             0  26206-26214 143rd Ave SE       Kent  WA 98042   
2      1963             0           857 170th Pl NE   Bellevue  WA 98008   
3      1976          1992         9105 170th Ave NE    Redmond  WA 98052   
4      1938          1994            522 NE 88th St    Seattle  WA 98115   

  country  
0     USA  
1     USA  
2     USA  
3     USA  
4     USA

In [38]:
house.describe()

price     bedrooms    bathrooms  sqft_living      sqft_lot  \
count    3421.000000  3421.000000  3421.000000  3421.000000  3.421000e+03   
mean   390913.274530     3.251681     1.972961  1822.543993  1.323065e+04   
std    137566.383127     0.874643     0.685133   680.343642  3.356439e+04   
min         0.000000     1.000000     0.750000   370.000000  6.380000e+02   
25%    287919.782600     3.000000     1.500000  1330.000000  5.000000e+03   
50%    390000.000000     3.000000     2.000000  1740.000000  7.500000e+03   
75%    499950.000000     4.000000     2.500000  2210.000000  1.009000e+04   
max    649950.000000     9.000000     6.250000  8020.000000  1.074218e+06   

            floors   waterfront         view    condition   sqft_above  \
count  3421.000000  3421.000000  3421.000000  3421.000000  3421.000000   
mean      1.439491     0.002923     0.117802     3.431453  1570.168956   
std       0.537885     0.053995     0.534080     0.658542   633.103923   
min       1.000000     0.000000     0.000000     1.000000   370.000000   
25%       1.000000     0.000000     0.000000     3.000000  1110.000000   
50%       1.000000     0.000000     0.000000     3.000000  1410.000000   
75%       2.000000     0.000000     0.000000     4.000000  1910.000000   
max       3.500000     1.000000     4.000000     5.000000  8020.000000   

       sqft_basement     yr_built  yr_renovated  
count    3421.000000  3421.000000   3421.000000  
mean      252.375037  1970.512716    832.157556  
std       391.020783    28.497013    983.919927  
min         0.000000  1900.000000      0.000000  
25%         0.000000  1952.000000      0.000000  
50%         0.000000  1973.000000      0.000000  
75%       500.000000  1994.000000   2000.000000  
max      2110.000000  2014.000000   2014.000000

In [29]:
### 설명변수(x)와 반응변수(y)
x = house[['bedrooms', 'bathrooms','yr_renovated', 'view', 'yr_built']]
#설명변수가 많아지면 accuracy가 증가한다.
# accuracy를 증가시키려면  가격에 영향을 많이 주는 변수를 잘 찾아양 한다.
y = house["price_cat"]

In [30]:
### Train & Test data
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.4)

In [31]:
### logistic regression
# multi_class = {'auto','ovr','multinomial'}, default= 'auto'
logistic = LogisticRegression(multi_class = 'ovr')
model = logistic.fit(x_train,y_train)

In [32]:
### 예측
y_pred = model.predict(x_test)
y_pred

array(['Class2', 'Class4', 'Class2', ..., 'Class2', 'Class2', 'Class2'],
      dtype=object)

In [33]:
### 교차표
pd.crosstab(y_test,y_pred)

col_0      Class2  Class3  Class4
price_cat                        
Class2        318     139      59
Class3        177     172     144
Class4         83     119     158

In [34]:
### accuracy
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

      Class2       0.55      0.62      0.58       516
      Class3       0.40      0.35      0.37       493
      Class4       0.44      0.44      0.44       360

    accuracy                           0.47      1369
   macro avg       0.46      0.47      0.46      1369
weighted avg       0.47      0.47      0.47      1369



In [35]:
# accuracy
model.score(x_test,y_test)

0.4733382030679328

설명변수 'bedrooms', 'bathrooms', 'sqft_living', 'view', 'condition'
0.48867786705624544

설명변수 'bedrooms', 'bathrooms','yr_renovated', 'view', 'yr_built'
0.4733382030679328